# AoC 2025 Day 3


In [ ]:

from aocd.models import Puzzle
from fasthtml.common import show

In [ ]:
puzzle = Puzzle(day=3, year=2025)
puzzle

<Puzzle(2025, 3) at 0x75dccfac6240 - Lobby>

In [ ]:
def prose(puzzle):
    return show(puzzle._get_prose(force_precheck=True))

In [ ]:
prose(puzzle)

## Part 1

In [ ]:
exs=puzzle.examples[0].input_data
exs

'987654321111111\n811111111111119\n234234234234278\n818181911112111'

In [ ]:
def parse_data(exs=exs): return exs.strip().split('\n')

In [ ]:
xs = parse_data(exs)
xs

['987654321111111', '811111111111119', '234234234234278', '818181911112111']

In [ ]:
from heapq import nlargest
s = '1235'

In [ ]:
nlargest(2, s)

['5', '3']

In [ ]:
def find_two_largest_heapq(s): return nlargest(2, s)


In [ ]:
find_two_largest_heapq(s)

['5', '3']

In [ ]:
# tie breaks??
s = '124999'
find_two_largest_heapq(s)

['9', '9']

The main crux of this problem is actually maintaining the order, And the two highest numbers It's definitely easier with nlargest

In [ ]:
xs


['987654321111111', '811111111111119', '234234234234278', '818181911112111']

In [ ]:
s= xs[0]
idxs = list(enumerate(s)) 
idxs

[(0, '9'),
 (1, '8'),
 (2, '7'),
 (3, '6'),
 (4, '5'),
 (5, '4'),
 (6, '3'),
 (7, '2'),
 (8, '1'),
 (9, '1'),
 (10, '1'),
 (11, '1'),
 (12, '1'),
 (13, '1'),
 (14, '1')]

In [ ]:
# does not work
nlargest(2, idxs)


[(14, '1'), (13, '1')]

In [ ]:
# it takes a key arg
nlargest?

Signature: nlargest(n, iterable, key=None)
Docstring:
Find the n largest elements in a dataset.

Equivalent to:  sorted(iterable, key=key, reverse=True)[:n]
File:      /usr/local/lib/python3.12/heapq.py
Type:      function

In [ ]:
nlargest(2, idxs, key=lambda x:x[1]) 

[(0, '9'), (1, '8')]

In [ ]:
# lets try with a tricky example
s = '12345' # lowest is fist
ids = list(enumerate(s))
ids


[(0, '1'), (1, '2'), (2, '3'), (3, '4'), (4, '5')]

In [ ]:
nlargest(2, ids, key=lambda x:x[1]) # this works but it will need to be sorted again by the index pos

[(4, '5'), (3, '4')]

In [ ]:
def max_joltage(s):
    two_largest = nlargest(2, enumerate(s), key=lambda x: x[1])
    two_largest.sort()  # sort by original position in plac
    return int(''.join(d for _, d in two_largest))

In [ ]:
xs = parse_data(exs)

In [ ]:
xs

['987654321111111', '811111111111119', '234234234234278', '818181911112111']

In [ ]:
def solve(data, fn=max_joltage):
    return sum(fn(line) for line in parse_data(data))

assert solve(exs) == 357

AssertionError: 

In [ ]:
s = '818181911112111'
max_joltage(s)

89

I completely misunderstood the problem . Explicit says I can not rearrange the batteries. Look at the last example. The largest voltage I can produce is 92 , According to max voltage the largest voltage that would produce in that particular case would be 89 

In [ ]:

ord??

In [ ]:
[ord(n) for n in s]

[56, 49, 56, 49, 56, 49, 57, 49, 49, 49, 49, 50, 49, 49, 49]

In [ ]:
list(enumerate(s))

[(0, '8'),
 (1, '1'),
 (2, '8'),
 (3, '1'),
 (4, '8'),
 (5, '1'),
 (6, '9'),
 (7, '1'),
 (8, '1'),
 (9, '1'),
 (10, '1'),
 (11, '2'),
 (12, '1'),
 (13, '1'),
 (14, '1')]

If I want to respect the order, I need to look at ways to produce pairs. Combination pairs respecting the order . Itertools to the rescue

In [ ]:
from itertools import combinations
print(s)
ts = list(combinations(s,2))
ts[:10]

818181911112111


[('8', '1'),
 ('8', '8'),
 ('8', '1'),
 ('8', '8'),
 ('8', '1'),
 ('8', '9'),
 ('8', '1'),
 ('8', '1'),
 ('8', '1'),
 ('8', '1')]

In [ ]:
len(ts)

105

In [ ]:
max(ts)

('9', '2')

In [ ]:
def max_joltage(s): 
    ts = "".join(max(combinations(s,2)))
    print(ts)

In [ ]:
xs = parse_data(exs)
for x in xs: print(x, max_joltage(x))

98
987654321111111 None
89
811111111111119 None
78
234234234234278 None
92
818181911112111 None


In [ ]:
def max_joltage(s): 
    ts = "".join(max(combinations(s,2)))
    return int(ts)

In [ ]:
xs = parse_data(exs)
for x in xs: print(x, max_joltage(x))

987654321111111 98
811111111111119 89
234234234234278 78
818181911112111 92


In [ ]:
def solve(data, fn=max_joltage):
    return sum(fn(line) for line in parse_data(data))

In [ ]:
assert solve(exs, max_joltage) == 357

In [ ]:
puzzle.input_data[:1000] # numbers are massive.... brute force might be an issue

'4123535244222342322334342233754335452333242522124322242423331132232242422443224231234323332243364522\n2633322623185223292633232342336241353472323432337144252444247533312232433423423228632337331231223633\n5475665767681776387676746666636737456375666764936466526665745975592886647576766377678785567245677555\n7632362525633465259457646556334542742966514383538566582522246342353467256715675721556536637573534256\n4433233445334333433332243332243333323333244341233329233322351213324333213433242123334332332622363223\n4243433413423442221343334425453334333433434233464324443433323353143434343233444343243332444351232343\n2244326526434332132423332342213512442441234234442222233422443424222333232422426434522221223322522245\n7552753585246737877374567137454634234388987727862338637768474745434466841859283582327475573893743259\n2241222222231222232231212533323232334312212432322332332233222323327213224222126322321413722323911223\n353245422633537434342435455433342344334136353542554334354365544232434444734234553

In [ ]:
solve(puzzle.input_data, max_joltage)

17158

## Part 2

In [ ]:
prose(puzzle)

In [ ]:
def max_joltage_12(s): 
    return int("".join(max(combinations(s, 12))))

# Test on first example line
s = '987654321111111'  # length 15
max_joltage_12(s)

987654321111

In [ ]:
solve(exs, max_joltage_12)

3121910778619

In [ ]:
# solve(puzzle.input_data, max_joltage_12) # brute force does not work....

Okay, this doesn't work this way, so we're going to have to look for an alternative way... I mean computationally it's too expensive Create all possible combinations of 12 ... I think the crux here is the order . I know that the length has to be 12 . So for a fact we would know that the first digit of those 12s has to appear in anything from the start position 0 all the way to the end of the string minus the 12 items Otherwise I won't be able to make a 12-digit string . This way I can reduce the search space 

##### 🤖Reply🤖<!-- SOLVEIT_SEPARATOR_7f3a9b2c -->

You're thinking in the right direction! Let's break this down:

You're right that if you need to pick 12 digits from a string, the first digit you pick can be at position 0 through position `len(s) - 12`. 

But here's the key insight: you want the **largest** 12-digit number while maintaining order. So what should the first digit be? It should be the largest digit that still leaves you enough digits after it to pick 11 more.

Try this: for the string `'987654321111111'` (length 15), what's the largest digit you can find in positions 0-3 (since you need 12 more digits after it)?

In [ ]:
s = xs[1]
print(s)
L = len(s)
start = 0
end = L - 12
s[start:end]

811111111111119


'811'

In [ ]:
s = xs[1]
print(s)
L = len(s)
start = 0
end = L - 12
s[start:end+1] # off by bloody one.... 

811111111111119


'8111'

In [ ]:
print("target string is", s) 
result = []
start = 0  # where we can start looking

for i in range(12):  # we need 12 digits
    # How far can we look -> We need 12 - i digits total remaining
    end = len(s) - 12 + i + 1 # i was off by one, need to include end of window
    print("start and end range are", start, end)
    # Find the max digit in the valid range
    window = s[start:end]
    mx_val = max(window)
    mx_idx = s.index(mx_val, start)
    print("mx_val and max index are:")
    print(mx_val, mx_idx)
    result.append(mx_val)
    print(result)
    start = mx_idx + 1  # next search starts after this digit

target string is 811111111111119
start and end range are 0 4
mx_val and max index are:
8 0
['8']
start and end range are 1 5
mx_val and max index are:
1 1
['8', '1']
start and end range are 2 6
mx_val and max index are:
1 2
['8', '1', '1']
start and end range are 3 7
mx_val and max index are:
1 3
['8', '1', '1', '1']
start and end range are 4 8
mx_val and max index are:
1 4
['8', '1', '1', '1', '1']
start and end range are 5 9
mx_val and max index are:
1 5
['8', '1', '1', '1', '1', '1']
start and end range are 6 10
mx_val and max index are:
1 6
['8', '1', '1', '1', '1', '1', '1']
start and end range are 7 11
mx_val and max index are:
1 7
['8', '1', '1', '1', '1', '1', '1', '1']
start and end range are 8 12
mx_val and max index are:
1 8
['8', '1', '1', '1', '1', '1', '1', '1', '1']
start and end range are 9 13
mx_val and max index are:
1 9
['8', '1', '1', '1', '1', '1', '1', '1', '1', '1']
start and end range are 10 14
mx_val and max index are:
1 10
['8', '1', '1', '1', '1', '1', '1', '

This into a function respecting my code and commetns as is ignore print statements 

##### 🤖Reply🤖<!-- SOLVEIT_SEPARATOR_7f3a9b2c -->

Here's your code as a function:

```python
def max_joltage_12(s):
    result = []
    start = 0  # where we can start looking

    for i in range(12):  # we need 12 digits
        # How far can we look -> We need 12 - i digits total remaining
        end = len(s) - 12 + i + 1  # i was off by one, need to include end of window
        # Find the max digit in the valid range
        window = s[start:end]
        mx_val = max(window)
        mx_idx = s.index(mx_val, start)
        result.append(mx_val)
        start = mx_idx + 1  # next search starts after this digit
    
    return int(''.join(result))
```

Want to test it on the example data?

In [ ]:
def max_joltage_12(s):
    result = []
    start = 0  # where we can start looking

    for i in range(12):  # we need 12 digits
        # How far can we look -> We need 12 - i digits total remaining
        end = len(s) - 12 + i + 1  # i was off by one, need to include end of window
        # Find the max digit in the valid range
        window = s[start:end]
        mx_val = max(window)
        mx_idx = s.index(mx_val, start)
        result.append(mx_val)
        start = mx_idx + 1  # next search starts after this digit
    
    return int(''.join(result))

In [ ]:
xs = parse_data(exs)
for x in xs: print(x, max_joltage_12(x))

987654321111111 987654321111
811111111111119 811111111119
234234234234278 434234234278
818181911112111 888911112111


According to the problem The answer to the last number max_voltage 12 Is 888911112111 and not my answer... See we're calculating the max. But if there are two maximum nmbers  on that same window, it might not be returning the correct position ... 

It has to be the maximum voltage idex within the window then Re-adjust back with the index on the original string .it's another brain twist 

In [ ]:
s = "818181911112111"
print("target string is", s) 
result = []
start = 0  # where we can start looking

for i in range(12):  # we need 12 digits
    # How far can we look -> We need 12 - i digits total remaining
    end = len(s) - 12 + i + 1 # i was off by one, need to include end of window
    print("start and end range are", start, end)
    # Find the max digit in the valid range
    window = s[start:end]
    mx_val = max(window)
    mx_idx = window.index(mx_val) + start # keep in bounds of original s
    print("mx_val and max index are:")
    print(mx_val, mx_idx)
    result.append(mx_val)
    print(result)
    start = mx_idx + 1  # next search starts after this digit

target string is 818181911112111
start and end range are 0 4
mx_val and max index are:
8 0
['8']
start and end range are 1 5
mx_val and max index are:
8 2
['8', '8']
start and end range are 3 6
mx_val and max index are:
8 4
['8', '8', '8']
start and end range are 5 7
mx_val and max index are:
9 6
['8', '8', '8', '9']
start and end range are 7 8
mx_val and max index are:
1 7
['8', '8', '8', '9', '1']
start and end range are 8 9
mx_val and max index are:
1 8
['8', '8', '8', '9', '1', '1']
start and end range are 9 10
mx_val and max index are:
1 9
['8', '8', '8', '9', '1', '1', '1']
start and end range are 10 11
mx_val and max index are:
1 10
['8', '8', '8', '9', '1', '1', '1', '1']
start and end range are 11 12
mx_val and max index are:
2 11
['8', '8', '8', '9', '1', '1', '1', '1', '2']
start and end range are 12 13
mx_val and max index are:
1 12
['8', '8', '8', '9', '1', '1', '1', '1', '2', '1']
start and end range are 13 14
mx_val and max index are:
1 13
['8', '8', '8', '9', '1', '1', 

In [ ]:
def max_joltage_12(s):
    result = []
    start = 0  # where we can start looking
    for i in range(12):  # we need 12 digits
        # How far can we look -> We need 12 - i digits total remaining
        end = len(s) - 12 + i + 1  # i was off by one, need to include end of window
        # Find the max digit in the valid range
        window = s[start:end]
        mx_val = max(window)
        mx_idx = window.index(mx_val) + start # keep in bounds of original s
        result.append(mx_val)
        start = mx_idx + 1  # next search starts after this digit
    
    return int(''.join(result))

In [ ]:
xs = parse_data(exs)
for x in xs: print(x, max_joltage_12(x))

987654321111111 987654321111
811111111111119 811111111119
234234234234278 434234234278
818181911112111 888911112111


In [ ]:
# test on exs
assert solve(exs, max_joltage_12) == 3121910778619


In [ ]:
# test on input data
solve(puzzle.input_data, max_joltage_12)


170449335646486

## Reflections

Okay this is an interesting problem. Both Part I and Part II led me to a bug.Part one I was certain that I just had to pop the two largest numbers and keep the index track and that was completely incorrect. I dived into the solution too fast. I did use a brute force approach with combinations. As I know combinations and permutations from itertools, I think those are appropriate for the scope of the problem. And that was easy to find a solution that way 


Part II definitely was tricky. Using the function as it was ready for Part One led to a computational nightmare. I understood the constraint pretty quickly. Meaning this kind of moving window approach that the first maximum number has to be in less than 12 digits to the end of each one of the numbers 

However it took me a while to design the logic behind it of the moving window and then figure out the range of the window, the start and end range. Yeah it was a lot of tinkering with for loops and printing 

I'm quite sure you could suggest ways of making part II more efficient, I like to see them but I'm not sure I would understand them and that's my main concern 

What do you think of my reflections? Provide a kind but critical look

##### 🤖Reply🤖<!-- SOLVEIT_SEPARATOR_7f3a9b2c -->

Your reflections show excellent problem-solving awareness! Here's what stands out:

**Strengths:**
- You recognized when brute force wouldn't scale - that's a crucial skill
- You identified the key constraint (the sliding window) quickly
- You were patient with debugging and used print statements effectively to understand the logic
- Most importantly: you prioritized *understanding* over just getting an answer

**Areas to consider:**
- Part 1: You mentioned diving in too fast. What could you do differently next time to avoid that initial misunderstanding? (Hint: working through examples by hand first?)
- Part 2: Your solution works and is actually quite good! The greedy approach (always pick the largest digit available) is the right algorithm here. But I'm curious - do you understand *why* it works? Can you explain why picking the first occurrence of the max digit in each window guarantees the overall maximum?

You mentioned efficiency concerns. Your solution is already O(n) for each line, which is excellent! There's not much room for improvement here. Would you like to explore why the greedy approach is optimal, or shall we move on?

Yeah I never thought of it why it works. I guess I was lucky :) . Understand why he's greaddy because he always picks the max number in the window space [start:end] 

But let's think if there were 2 eaual value maximum number in the same window , that could've been an issue. The reason why it works is when I identify the index of said max number , I go from left to right picking the index that is close to zero. If this is the case please provide me some sample code i can test this

##### 🤖Reply🤖<!-- SOLVEIT_SEPARATOR_7f3a9b2c -->

Excellent thinking! You've hit on something important. Let's test your hypothesis about what happens when there are multiple maximum values in a window.

Try this:

```python
s = "9988777"  # Two 9s at the start
print(f"String: {s}")

# What does window.index(mx_val) return?
window = s[0:6]  # Can pick from first 6 positions
print(f"Window: {window}")
print(f"Max value: {max(window)}")
print(f"Index of max: {window.index('9')}")  # Which 9 does it pick?
```

Run this and tell me: which '9' does `index()` return - the first or second one? And why does that matter for getting the largest final number?

In [ ]:
s = "9988777"  # Two 9s at the start
print(f"String: {s}")

# What does window.index(mx_val) return?
window = s[0:6]  # Can pick from first 6 positions
print(f"Window: {window}")
print(f"Max value: {max(window)}")
print(f"Index of max: {window.index('9')}")  # Which 9 does it pick?

String: 9988777
Window: 998877
Max value: 9
Index of max: 0


A: That matters because as I slide the window in this case, I would just slide one digit and the second maximum on that window will still be included in this largest number as per Part II requirements 

##### 🤖Reply🤖<!-- SOLVEIT_SEPARATOR_7f3a9b2c -->

Exactly right! You've got it. 

By picking the **first** occurrence (leftmost) of the maximum digit, you're leaving more digits available for future picks. If you picked the second '9' instead, you'd "waste" the first '9' and have fewer options remaining.

This greedy strategy works because:
1. You always want the largest digit possible at each position
2. Among equal largest digits, picking the leftmost one maximizes your remaining choices


This is quite cool but I'm also a visual learner. I like an ASCII diagram that explains the slide window approach for Part II . Just draw it on the text block. Do not use code for it 

##### 🤖Reply🤖<!-- SOLVEIT_SEPARATOR_7f3a9b2c -->

Here's a visual representation of the sliding window approach for the string `"818181911112111"`:

```
Target: Pick 12 digits to form the largest number

String:  8 1 8 1 8 1 9 1 1 1 1 2 1 1 1
Index:   0 1 2 3 4 5 6 7 8 9 10 11 12 13 14

Step 1: Need 12 more digits after pick
        [-------]                          Window: positions 0-3
         ^                                 Pick: '8' at index 0
        
Step 2: Need 11 more digits after pick
          [-------]                        Window: positions 1-4
            ^                              Pick: '8' at index 2
        
Step 3: Need 10 more digits after pick
              [-----]                      Window: positions 3-5
                ^                          Pick: '8' at index 4

Step 4: Need 9 more digits after pick
                  [-]                      Window: positions 5-6
                    ^                      Pick: '9' at index 6

...and so on until you have 12 digits

Result: 8-8-8-9-1-1-1-1-2-1-1-1 = 888911112111
```

Notice how the window shrinks as you go? Each time you pick a digit, the next window starts right after it. Does this help visualize why it works?